python/pyarrow/dataset.py

```diff
@@ -53,3 +57,236 @@
     SchemaPartitionScheme,
     TreeDataSource,
 )
+
+
+def partitioning(field_names=None, flavor=None):
+    """
+    Specify a partitioning scheme.
+
+    The supported schemes include:
+
+    - "SchemaPartitioning": this scheme expects one segment in the file path
+      for each field in the specified schema (all fields are required to be
+      present). For example given schema<year:int16, month:int8> the path
+      "/2009/11" would be parsed to ("year"_ == 2009 and "month"_ == 11).
+    - "HivePartitioning": a scheme for "/$key=$value/" nested directories as
+      found in Apache Hive. This is a multi-level, directory based partitionin
+      scheme with all data files stored in the leaf directories. Data is
+      partitioned by static values of a particular column in the schema.
+      Partition keys are represented in the form $key=$value in directory
+      names. Field order is ignored, as are missing or unrecognized field
+      names.
+      For example, given schema<year:int16, month:int8, day:int8>, a possible
+      path would be "/year=2009/month=11/day=15".
+
+    Parameters
+    ----------
+    field_names : pyarrow.Schema or list of str
+        The schema that describes the partitions present in the file path. If
+        a list of strings (field names) is passed, the schema's types are
+        inferred from the file paths (only valid for SchemaPartitioning).
+    flavor : str, default None
+        The default is SchemaPartitioning. Specify ``flavor="hive"`` for
+        a HivePartitioning.
+
+    Returns
+    -------
+    PartitionScheme or PartitionDiscovery
+
+    Examples
+    --------
+
+    Specify the Schema for paths like "/2009/11":
+
+    >>> partitioning(pa.schema([("year", pa.int16()), ("month", pa.int8())]))
+
+    or let the types be inferred by only specifying the field names:
+
+    >>> partitioning(["year", "month"])
```

Comment:
> It might be handy to include the types which would be inferred in this case 

Good idea. Just explain this in the text? Because I don't think there is an easy way to "show" this with code, right? Because then the partitioning needs to be tied to an actual dataset.